Add README and boilerplate projects for Microsoft.DotNet.Framework.Sdk

natemcmaster opened the pull request:
Part of #2704 

This lays out an empty project and a detailed README for a new package I'm starting, Microsoft.DotNet.Framework.Sdk.

Note: This is targeting a feature branch. Expect several more PRs before this feature is ready to merge to master.

dagood commented on src/Microsoft.DotNet.SharedFramework.Sdk/README.md:
```diff
@@ -0,0 +1,80 @@
+Projects which need to create a .NET Core shared framework can use this SDK to generate some of the required content. Shared frameworks
+generally require at least two kinds of packages, so the recommended usage is to create two projects, one for the targeting pack and another for the runtime packs. 
+Both projects can use a custom MSBuild SDK, "Microsoft.DotNet.SharedFramework.Sdk" and set some well-known properties which will inform the
+custom SDK how to build a shared framework.
```
In my experience, 80 columns is expected if you're going to wrap. E.g. in the normal GitHub review panel the current wrapping causes the text to become jagged. (And many people do like markdown wrapping because it *generally* makes better diffs and works better in review tools that can't do their own wrapping.)

dagood commented on src/Microsoft.DotNet.SharedFramework.Sdk/README.md:
```diff
@@ -0,0 +1,80 @@
+# Microsoft.DotNet.SharedFramework.Sdk
+
+Projects which need to create a .NET Core shared framework can use this SDK to generate some of the required content. Shared frameworks
+generally require at least two kinds of packages, so the recommended usage is to create two projects, one for the targeting pack and another for the runtime packs. 
```
I don't really understand the reason to have multiple projects, I brought this up in https://github.com/dotnet/arcade/issues/2704#issuecomment-490641216:

> 4. On a similar note... I don't see a good reason to support `PlatformPackageType` but not (say) `PlatformPackageTypes`. (In any case, a way to build multiple packs from one project that contains all the info.) I don't think it's unreasonable to have one project for each pack type but listing some justification would be nice.

For the sake of this PR I'd be fine just dropping this sentence from the readme.

natemcmaster replied:
Primary reason for multiple project files is that I’m not aware of a way to make NuGet produce multiple NuGet packages per project, or vary the project references on a property (other than TargetFramework).

dagood replied:
I know very little about the NuGet targets, sounds reasonable to me.

dagood commented on src/Microsoft.DotNet.SharedFramework.Sdk/README.md:
```diff
@@ -0,0 +1,80 @@
+  <PropertyGroup>
+     <!-- Required properties -->
+    <TargetFramework>netcoreapp3.0</TargetFramework>
+    <RuntimeIdentifier>win-x64</RuntimeIdentifier>
```
I still don't understand some things about these XML files, but it appears to be the same as in the issue so I'm just noting that in general--don't know if we need to address it now.

natemcmaster replied:
One of my goals is to use as many features from Microsoft.NET.Sdk as possible. RuntimeIdentifier is a required property when you want NuGet to resolve rid-specific assets from packages.

dagood replied:
Following up on this, then:

> > RuntimeIdentifier
> 
> Didn't mean to imply a separate project per file. In AspNetCore, we always set RuntimeIdentifier and infer it from the host machine (unless it is explicitly set by build script parameters).
> 
> I think RuntimeIdentifier should still be required because there is no such thing as an "AnyCPU/AnyOS" shared framework. `RuntimeIdentifiers` could additionally be set, but not sure if requiring it is important.

Why is `win-x64` in particular in the project file? Is just some value required and this is arbitrarily the one that's stuck in the project as a placeholder? It sticks out quite a bit to me.

natemcmaster replied:
> Why is win-x64 in particular in the project file?

Just meant to be a sample. In the core-setup repo, I would expect the project file to use the property like you already do: https://github.com/dotnet/core-setup/blob/98fe009648f423dc783645714fd2e473eaf37f66/src/pkg/packaging-tools/sharedFramework/sharedFramework.csproj
```xml
<RuntimeIdentifier>$(PackageRID)</RuntimeIdentifier>
```

dagood replied:
Oh! Thanks, that makes sense. If there's some property like that provided by Arcade normally, that would be ideal in the example IMO. Not sure if there is anything like that though.

natemcmaster replied:
I don't think there is a standard property across the repos. Wish there were, but for legacy reasons, changing that seems out of scope for now. Corefx uses "$(OSGroup)" and "$(ArchGroup)". AspNetCore uses "$(TargetRuntimeIdentifier)". I'm not actually sure how PackageID gets set in core-setup.

natemcmaster commented:
I started building this out and realize now this could be used for NETStandard.Library.  I think a better name for this is Microsoft.DotNet.Framework.Sdk or Microsoft.DotNet.Platform.Sdk, since netstandard isn’t a shared runtime. Thoughts?

dagood commented:
> I started building this out and realize now this could be used for NETStandard.Library. I think a better name for this is Microsoft.DotNet.Framework.Sdk or Microsoft.DotNet.Platform.Sdk, since netstandard isn’t a shared runtime. Thoughts?

Agreed that these shouldn't go under a banner of "shared framework", even without considering the netstandard case. In my mind, targeting and runtime packs are pretty separate from the idea of a shared framework because they target specific scenarios. I haven't been working with shared frameworks for very long so my terminology may be strange, though.

I've typically called the targeting and runtime (and AppHost) packs "framework packs", FWIW. Not sure of the top of my head what to call something that creates shared frameworks and framework packs.

dagood commented:
> Microsoft.DotNet.Framework.Sdk or Microsoft.DotNet.Platform.Sdk

I failed to mention--either sounds fine to me, and of the two I think `Microsoft.DotNet.Framework.Sdk` aligns a little better with what I've been calling these things.

natemcmaster commented:
Cool, I think Microsoft.DotNet.Framework.Sdk aligns better too. I'll update the readme and project files.

natemcmaster commented:
Updated. Changes:

* Microsoft.DotNet.{SharedFramework => Framework}.Sdk
* Renamed SharedFrameworkName to FrameworkName
* Renamed PlatformPackageType to FrameworkPackageType
* Cleaned up readme and samples, including clarifying on the point about RuntimeIdentifier

natemcmaster commented:
Applied feedback and merged. I'll get started on implementing some of these targets and tasks next week.
